code/modules/surgery/organs/stomach/external/wings.dm

```diff
@@ -0,0 +1,203 @@
+	if(H.movement_type & FLYING)
+		if(!CanFly(H))
+			ToggleFlight(H)
+			return FALSE
+		return TRUE
+	else
+		return FALSE
+
+///Check if we can still fly
+/obj/item/organ/external/wings/functional/proc/CanFly(mob/living/carbon/human/H)
+	if(H.stat || H.body_position == LYING_DOWN)
+		return FALSE
+	if(H.wear_suit && ((H.wear_suit.flags_inv & HIDEJUMPSUIT) && (!H.wear_suit.species_exception || !is_type_in_list(src, H.wear_suit.species_exception)))) //Jumpsuits have tail holes, so it makes sense they have wing holes too
+		to_chat(H, span_warning("Your suit blocks your wings from extending!"))
+		return FALSE
+	var/turf/T = get_turf(H)
+	if(!T)
+		return FALSE
+
+	var/datum/gas_mixture/environment = T.return_air()
+	if(environment && !(environment.return_pressure() > 30))
+		to_chat(H, span_warning("The atmosphere is too thin for you to fly!"))
+		return FALSE
+	else
+		return TRUE
+
+///Slipping but in the air?
+/obj/item/organ/external/wings/functional/proc/flyslip(mob/living/carbon/human/H)
+	var/obj/buckled_obj
+	if(H.buckled)
+		buckled_obj = H.buckled
+
+	to_chat(H, span_notice("Your wings spazz out and launch you!"))
+
+	playsound(H.loc, 'sound/misc/slip.ogg', 50, TRUE, -3)
+
+	for(var/obj/item/I in H.held_items)
+		H.accident(I)
+
+	var/olddir = H.dir
+
+	H.stop_pulling()
+	if(buckled_obj)
+		buckled_obj.unbuckle_mob(H)
+		step(buckled_obj, olddir)
+	else
+		new /datum/forced_movement(H, get_ranged_target_turf(H, olddir, 4), 1, FALSE, CALLBACK(H, /mob/living/carbon/.proc/spin, 1, 1))
+	return TRUE
+
+///UNSAFE PROC, should only be called through the Activate or other sources that check for CanFly
+/obj/item/organ/external/wings/functional/proc/ToggleFlight(mob/living/carbon/human/H)
+	if(!HAS_TRAIT_FROM(H, TRAIT_MOVE_FLYING, SPECIES_FLIGHT_TRAIT))
+		H.physiology.stun_mod *= 2
+		ADD_TRAIT(H, TRAIT_NO_FLOATING_ANIM, SPECIES_FLIGHT_TRAIT)
+		ADD_TRAIT(H, TRAIT_MOVE_FLYING, SPECIES_FLIGHT_TRAIT)
+		passtable_on(H, SPECIES_TRAIT)
+		OpenWings()
+	else
+		H.physiology.stun_mod *= 0.5
+		REMOVE_TRAIT(H, TRAIT_NO_FLOATING_ANIM, SPECIES_FLIGHT_TRAIT)
+		REMOVE_TRAIT(H, TRAIT_MOVE_FLYING, SPECIES_FLIGHT_TRAIT)
+		passtable_off(H, SPECIES_TRAIT)
+		CloseWings()
+
+///SPREAD OUR WINGS AND FLLLLLYYYYYY
+/obj/item/organ/external/wings/functional/proc/OpenWings()
+	preference = wings_open_preference
+	wings_open = TRUE
+
+	generate_icon_cache()
+	owner.update_body_parts()
+
+///close our wings
+/obj/item/organ/external/wings/functional/proc/CloseWings()
+	preference = wings_closed_preference
+	wings_open = FALSE
+
+	generate_icon_cache()
+	owner.update_body_parts()
+	if(isturf(owner?.loc))
+		var/turf/T = loc
+		T.Entered(src, NONE)
+
+///hud action for starting and stopping flight
+/datum/action/innate/flight
+	name = "Toggle Flight"
+	check_flags = AB_CHECK_CONSCIOUS|AB_CHECK_IMMOBILE
+	icon_icon = 'icons/mob/actions/actions_items.dmi'
+	button_icon_state = "flight"
+
+/datum/action/innate/flight/Activate()
+	var/mob/living/carbon/human/H = owner
```

Comment:
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names